### PR TITLE
[MXNET-1182] Predictor example

### PR DESCRIPTION
**Pull request by lanking520**
## Description ##
Adding draft example using predictor.
Remaining tasks:

- [x] Add README
- [x] Add script to run

@piyushghai @andrewfayres @nswamy 
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by anirudhacharya** [comment]
@mxnet-label-bot add [pr-work-in-progress]

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+import java.awt.Graphics2D;
+import java.awt.image.BufferedImage;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class PredictorExample {
```
Comment (piyushghai) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java]: I think this class should be renamed to ImageClassificationExample 
And in comments we can write that this is an example code which uses Predictor API.

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+            }
+            row += 1;
+        }
+        buf.flush();
+        return result;
+    }
+
+    private static String printMaximumClass(float[] probabilities,
+                                            String modelPathPrefix) throws IOException {
+        String synsetFilePath = modelPathPrefix.substring(0,
```
Comment (piyushghai) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java]: We should not take the synsetFilePath from the modelPathPrefix.  
This enforces a hard dependency on the synset file to be present in the model folder...

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        // Prepare data
+        BufferedImage img = loadIamgeFromFile(inst.inputImagePath);
+
+        img = reshapeImage(img, 224, 224);
+        // predict
+        float[][] result = predictor.predict(new float[][]{imagePreprocess(img)});
+        try {
+            System.out.println("Predict with Float input");
+            System.out.println(printMaximumClass(result[0], inst.modelPathPrefix));
+        } catch (Exception e) {
```
Comment (piyushghai) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java]: catch the particular IOException here.

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                imagePreprocess(img),
+                new Shape(new int[]{1, 3, 224, 224}),
+                Context.cpu());
+        List<NDArray> ndList = new ArrayList<>();
+        ndList.add(nd);
+        List<NDArray> ndResult = predictor.predictWithNDArray(ndList);
+        try {
+            System.out.println("Predict with NDArray");
+            System.out.println(printMaximumClass(ndResult.get(0).toArray(), inst.modelPathPrefix));
+        } catch (Exception e) {
```
Comment (piyushghai) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java]: Same here. IO Ex

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/README.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+```
+
+**Notes**:
+* These are relative paths to this script.
+* You may need to run `chmod +x run_predictor_java_example.sh` before running this script.
+
+The example should give expected output as shown below:
+```
+Probability : 0.30337515 Class : n02123159 tiger cat
+Probability : 0.30337515 Class : n02123159 tiger cat
```
Comment (piyushghai) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/README.md]: Your output also has print statements about : "Predict with NDArray"

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java (hunk 2 above)
Comment (lanking520) [reply scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java]: I agree with you with this part, but they usually ship together in the same location. Most of the examples, SSD/Image classifier do this...

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java (hunk 1 above)
Comment (lanking520) [reply scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java]: As we will have an ImageClassification Example once we introduce Image Classifier...

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/README.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+1. [Prerequisites](#prerequisites)
+2. [Download artifacts](#download-artifacts)
+3. [Setup datapath and parameters](#setup-datapath-and-parameters)
+4. [Run the image inference example](#run-the-image-inference-example)
+
+## Prerequisites
+
+1. MXNet
+2. MXNet Scala Package
```
Comment (piyushghai) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/README.md]: Is this correct ?

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/README.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+## Contents
+
+1. [Prerequisites](#prerequisites)
+2. [Download artifacts](#download-artifacts)
+3. [Setup datapath and parameters](#setup-datapath-and-parameters)
+4. [Run the image inference example](#run-the-image-inference-example)
+
+## Prerequisites
+
+1. MXNet
```
Comment (piyushghai) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/README.md]: Can you also provide a link to the install page of MXNet ?

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/README.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+```bash
+bash run_predictor_java_example.sh ../models/resnet-18/resnet-18 ../images/kitten.jpg
+```
+
+**Notes**:
+* These are relative paths to this script.
+* You may need to run `chmod +x run_predictor_java_example.sh` before running this script.
+
+The example should give expected output as shown below:
```
Comment (piyushghai) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/README.md]: The example should give an output similar to the one shown below:

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/README.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+In this example, you will learn how to use Java Inference API to 
+build and run pre-trained Resnet 18 model.
+
+## Contents
+
+1. [Prerequisites](#prerequisites)
+2. [Download artifacts](#download-artifacts)
+3. [Setup datapath and parameters](#setup-datapath-and-parameters)
+4. [Run the image inference example](#run-the-image-inference-example)
```
Comment (piyushghai) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/README.md]: Run the image classification example

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+            System.out.println("Predict with Float input");
+            System.out.println(printMaximumClass(result[0], inst.modelPathPrefix));
+        } catch (IOException e) {
+            System.err.println(e);
+        }
+        // predict with NDArray
+        NDArray nd = new NDArray(
+                imagePreprocess(img),
+                new Shape(new int[]{1, 3, 224, 224}),
+                Context.cpu());
```
Comment (piyushghai) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java]: Quick question : Should this context be cpu ? Or determined based on the environment variable you have ?

**Comment by kalyc** [comment]
Thanks for your contribution @lanking520 
Could you re-trigger CI with empty commit?

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+    private static BufferedImage reshapeImage(BufferedImage buf, int newWidth, int newHeight) {
+        BufferedImage resizedImage = new BufferedImage(newWidth, newHeight, BufferedImage.TYPE_INT_RGB);
+        Graphics2D g = resizedImage.createGraphics();
+        g.drawImage(buf, 0, 0, newWidth, newHeight, null);
+        g.dispose();
+        return resizedImage;
+    }
+
+    private static float[] imagePreprocess(BufferedImage buf) {
```
Comment (andrewfayres) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java]: Same as above.

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+        BufferedImage buf = null;
+        try {
+            buf = ImageIO.read(new File(inputImagePath));
+        } catch (IOException e) {
+            System.err.println(e);
+        }
+        return buf;
+    }
+
+    private static BufferedImage reshapeImage(BufferedImage buf, int newWidth, int newHeight) {
```
Comment (andrewfayres) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java]: Same as above.

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+
+public class PredictorExample {
+    @Option(name = "--model-path-prefix", usage = "input model directory and prefix of the model")
+    private String modelPathPrefix = "/model/ssd_resnet50_512";
+    @Option(name = "--input-image", usage = "the input image")
+    private String inputImagePath = "/images/dog.jpg";
+
+    final static Logger logger = LoggerFactory.getLogger(PredictorExample.class);
+
+    private static BufferedImage loadIamgeFromFile(String inputImagePath) {
```
Comment (andrewfayres) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java]: This same logic is in ObjectDetector. Is there any reason we aren't just reusing those methods to make this example simpler?

Comment (lanking520) [reply scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/PredictorExample.java]: The reason to do this is show how users can build their own preprocessing stages with Predictor API as well as there is no single method that can bring me a float array unless I do an NDArray extraction to get a float array

File: scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/README.md [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+In this example, you will learn how to use Java Inference API to 
+build and run pre-trained Resnet 18 model.
+
+## Contents
+
+1. [Prerequisites](#prerequisites)
+2. [Download artifacts](#download-artifacts)
+3. [Setup datapath and parameters](#setup-datapath-and-parameters)
+4. [Run the image classifier example](#run-the-image-inference-example)
```
Comment (nswamy) [inline scala-package/examples/src/main/java/org/apache/mxnetexamples/javaapi/infer/predictor/README.md]: Why Image Classification in Prediction? Is it possible to use another example even may be a simple Linear Regression model?

**Pull request merged**